proxy/context.go

```diff
@@ -0,0 +1,178 @@
+	deprecatedServed      bool
+	servedWithResponse    bool // to support the deprecated way independently
+	pathParams            map[string]string
+	stateBag              map[string]interface{}
+	originalRequest       *http.Request
+	originalResponse      *http.Response
+	outgoingHost          string
+	debugFilterPanics     []interface{}
+	outgoingDebugRequest  *http.Request
+	incomingDebugResponse *http.Response
+	startServe            time.Time
+}
+
+func defaultBody() io.ReadCloser {
+	return ioutil.NopCloser(&bytes.Buffer{})
+}
+
+func defaultResponse(r *http.Request) *http.Response {
+	return &http.Response{
+		StatusCode: http.StatusNotFound,
+		Header:     make(http.Header),
+		Body:       defaultBody(),
+		Request:    r,
+	}
+}
+
+func cloneURL(u *url.URL) *url.URL {
+	uc := *u
+	return &uc
+}
+
+func cloneRequestMetadata(r *http.Request) *http.Request {
+	return &http.Request{
+		Method:           r.Method,
+		URL:              cloneURL(r.URL),
+		Proto:            r.Proto,
+		ProtoMajor:       r.ProtoMajor,
+		ProtoMinor:       r.ProtoMinor,
+		Header:           cloneHeader(r.Header),
+		Body:             defaultBody(),
+		ContentLength:    r.ContentLength,
+		TransferEncoding: r.TransferEncoding,
+		Close:            r.Close,
+		Host:             r.Host,
+		RemoteAddr:       r.RemoteAddr,
+		RequestURI:       r.RequestURI,
+		TLS:              r.TLS,
+	}
+}
+
+func cloneResponseMetadata(r *http.Response) *http.Response {
+	return &http.Response{
+		Status:           r.Status,
+		StatusCode:       r.StatusCode,
+		Proto:            r.Proto,
+		ProtoMajor:       r.ProtoMajor,
+		ProtoMinor:       r.ProtoMinor,
+		Header:           cloneHeader(r.Header),
+		Body:             defaultBody(),
+		ContentLength:    r.ContentLength,
+		TransferEncoding: r.TransferEncoding,
+		Close:            r.Close,
+		Request:          r.Request,
+		TLS:              r.TLS,
+	}
+}
+
+func newContext(w http.ResponseWriter, r *http.Request, preserveOriginal bool) *context {
+	c := &context{
+		responseWriter: w,
+		request:        r,
+		stateBag:       make(map[string]interface{}),
+		outgoingHost:   r.Host,
+	}
+
+	if preserveOriginal {
+		c.originalRequest = cloneRequestMetadata(r)
+	}
+
+	return c
+}
+
+func (c *context) applyRoute(route *routing.Route, params map[string]string, preserveHost bool) {
+	c.route = route
+	if preserveHost {
+		c.outgoingHost = c.request.Host
+	} else {
+		c.outgoingHost = route.Host
+	}
+
+	c.pathParams = params
+}
+
+func (c *context) ensureDefaultResponse() {
+	if c.response == nil {
+		c.response = defaultResponse(c.request)
+		return
+	}
+
+	if c.response.Header == nil {
+		c.response.Header = make(http.Header)
+	}
+
+	if c.response.Body == nil {
+		c.response.Body = defaultBody()
+	}
+}
+
+func (c *context) deprecatedShunted() bool {
+	return c.deprecatedServed
+}
+
+func (c *context) shunted() bool {
+	return c.servedWithResponse
+}
+
+func (c *context) isShuntRoute() bool {
+	return c.route.Shunt
```

Comment:
i'd rather remove the function now and call c.route.Shunt directly. I am not too much in favor of starting a full preventive style for package internal functions in case of this piece of code. What do you think?